packages/cli/src/tools/releaseChecker/getLatestRelease.js

```diff
@@ -0,0 +1,157 @@
+/**
+ * @flow
+ */
+import path from 'path';
+import https from 'https';
+import semver from 'semver';
+import fs from 'fs';
+import logger from '../logger';
+
+export type Release = {
+  tag_name: string,
+  html_url: string,
+  draft: boolean,
+  prerelease: boolean,
+  created_at: string,
+  published_at: string,
+  body: string,
+};
+
+/**
+ * Checks via GitHub API if there is a newer stable React Native release and,
+ * if it exists, returns the release data.
+ *
+ * If the latest release is not newer or if it's a prerelease, the function
+ * will return null.
+ */
+export default (async function(currentVersion: string) {
+  logger.debug('Checking for a newer version of React Native');
+  try {
+    logger.debug(`Current version: ${currentVersion}`);
+
+    const cache = cacheManager.get();
+    let eTag;
+    if (cache) {
+      eTag = cache.eTag;
+      logger.debug(
+        `Cached release version: ${semver.coerce(cache.release.tag_name)}`,
+      );
+    }
+
+    logger.debug('Checking for newer releases on GitHub');
+    const response: Response = await getLatestRelease(eTag);
+
+    let latestRelease;
+    if (
+      response.statusCode === 200 &&
+      response.release &&
+      !response.release.prerelease
+    ) {
+      latestRelease = response.release;
+      logger.debug(
+        `Remote release version: ${semver.coerce(latestRelease.tag_name)}`,
+      );
+      // Update the cache.
+      cacheManager.set({
+        release: latestRelease,
+        eTag: response.eTag,
+      });
+    } else {
+      // Response status is 304, meaning that our cached release data
+      // is the most recent release available.
+      logger.debug('Cache is up-to-date with GitHub releases');
+      // $FlowFixMe: We know at this point that cache is not null
+      latestRelease = cache.release;
+    }
+
+    const latestVersion = semver.coerce(latestRelease.tag_name);
+    logger.debug(`Latest stable release: ${latestVersion}`);
+    if (semver.compare(latestVersion, currentVersion) === 1) {
+      return latestRelease;
+    } else {
+      return null;
+    }
+  } catch (e) {
+    logger.debug(
+      'Something went wrong with remote version checking, moving on',
+    );
+    return null;
+  }
+});
+
+type Headers = {
+  'User-Agent': string,
+  [header: string]: string,
+};
+
+type Response = {
+  eTag: string,
+  release: ?Release,
+  statusCode: number,
+};
+
+function getLatestRelease(eTag: ?string) {
+  return new Promise<Response>((resolve, reject) => {
+    const options = {
+      hostname: 'api.github.com',
+      path: '/repos/facebook/react-native/releases/latest',
```

Comment:
We cannot check latest from RN github for the same reasons you mentioned in OP. We can either:
1) check from npm (not ideal, because rn-diff-purge is usually updated with a delay)
2) check from rn-diff-purge – this way we can be sure that running `yarn react-native upgrade` will actually have something to fetch